### PR TITLE
Team 2 agent MVP

### PR DESCRIPTION
**Pull request by Eirikalb**
# Summary
Agent 2, added trust system, will extend this to have a coherent strategy based on our classification of agents as altruististic/greedy or neutral as well as our overall trust in them based. Currently trust is only updated in the gifting session.

- Foraging: 
  - Decide whether to hunt or fish based off of estimated number of hunters in next turn -Baseclient implementation for sharing foraging information

- Tax: 
  - Pay tax unless critical

- Giving and taking Gifts: 
  - Request when critical 
  - Opinion on other teams depends on trustworthiness level which is decided from inofrmation shared and gifts given 
  - Give gifts depending on trustworthiness level

- Commonpool allocation: 
   - estimate the threshold if not given 
   - categorise agents as altruists,fair sharers or free riders depending on their contributions 
   - give average amount that other agents are contributing 
   - special case when pool is struggling or there are too many free riders, contribute much more 
   - special case when the pool has a surplus according to our estimated threshold, contribute less 
   - take from common pool when critical

- IIFO: 
 - Assign each agent a parameter based on average vulnerability to disasters based off location 
 - predict disaster from probability distributiobbs of relevant parts

IIGO: Baseclient Implementation
## Additional Information

Auxiliary information goes here.

## Test Plan

- Very limited testing, will try to add lots more for the 9th

### CODE REVIEW

**Comment by QFSW** [comment]
# ⚠️Warning! This diff contains changes in protected folders

File: internal/clients/team2/commonpool.go [hunk 1]
```diff
@@ -0,0 +1,217 @@
...
+		if c.resourceLevelHistory[turn] >= c.resourceLevelHistory[turn-1] { //no resources taken by disaster
+			if disasterBasedAdjustment > 5 {
+				disasterBasedAdjustment -= 5
+			}
+		}
+		//disaster took our resources
+		disasterBasedAdjustment += 5
+	}
+	//change factor by if next mag > or < prev mag
+	return shared.Resources(float64(baseThreshold)*(magnitudePrediction/sampleMeanM)*(1+ourVulnerability)+disasterBasedAdjustment) + costOfLiving
```
Comment (lhl2617) [inline internal/clients/team2/commonpool.go:141]: ~vibe~ div0 check

File: internal/clients/team2/disaster.go [hunk 2]
```diff
@@ -0,0 +1,195 @@
...
+		}
+		overlapArchipelagoOutline := Outline{
+			Left:   GetMinMaxCoordinate(Max, shiftedArchipelagoOutline.Left, archipelagoGeography.XMin),
+			Right:  GetMinMaxCoordinate(Min, shiftedArchipelagoOutline.Right, archipelagoGeography.XMax),
+			Bottom: GetMinMaxCoordinate(Max, shiftedArchipelagoOutline.Bottom, archipelagoGeography.YMin),
+			Top:    GetMinMaxCoordinate(Min, shiftedArchipelagoOutline.Top, archipelagoGeography.YMax),
+		}
+
+		areaOfOverlap := (overlapArchipelagoOutline.Right - overlapArchipelagoOutline.Left) * (overlapArchipelagoOutline.Top - overlapArchipelagoOutline.Bottom)
+		islandDVPs[islandID] = areaOfOverlap / areaOfArchipelago
```
Comment (lhl2617) [inline internal/clients/team2/disaster.go]: ~vibe~ div0 check

File: internal/clients/team2/disaster.go [hunk 3]
```diff
@@ -0,0 +1,195 @@
...
+		Y: archipelagoGeography.YMin + (archipelagoGeography.YMax-archipelagoGeography.YMin)/2,
+	}
+	return archipelagoCentre
+}
+
+// GetTimeRemainingPrediction returns a prediction about the time remaining until the next disaster and the sample mean
+// of the RV X. The prediction is 1/sample mean of the Bernoulli RV, minus the turns since the last disaster.
+func GetTimeRemainingPrediction(c *client, totalTurns float64) (float64, int) {
+	totalDisasters := float64(len(c.disasterHistory))
+	sampleMeanX := totalDisasters / totalTurns
```
Comment (lhl2617) [inline internal/clients/team2/disaster.go:138]: multiple ~vibe~ div0 checks required here

File: internal/clients/team2/commonpool.go (hunk 1 above)
Comment (Eirikalb) [reply internal/clients/team2/commonpool.go:141]: Big vibe

Comment (Eirikalb) [reply internal/clients/team2/commonpool.go:141]: Fixed above

File: internal/clients/team2/disaster.go (hunk 3 above)
Comment (Eirikalb) [reply internal/clients/team2/disaster.go:138]: totalturns is always larger than 1

**Review by NGovani — APPROVED** [review]

**Pull request merged**